src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/ProjectKey.cs

```diff
@@ -3,41 +3,36 @@
 
 using System;
 using System.Diagnostics;
+using System.IO;
 using Microsoft.AspNetCore.Razor;
+using Microsoft.AspNetCore.Razor.Utilities;
 
 namespace Microsoft.CodeAnalysis.Razor.ProjectSystem;
 
 /// <summary>
 /// A very light wrapper around a file path, used to ensure consistency across the code base for what constitutes the unique
 /// identifier for a project.
 /// </summary>
-internal sealed class ProjectKey
+internal sealed class ProjectKey : IEquatable<ProjectKey>
 {
-    // ProjectKey represents the full path to the project file name. All creation logic is here in one place
-    // to ensure this is consistent.
-    public static ProjectKey From(HostProject hostProject) => new(hostProject.FilePath);
-    public static ProjectKey From(IProjectSnapshot project) => new(project.FilePath);
-    public static ProjectKey? From(Project project) => project.FilePath is null ? null : new(project.FilePath);
-    public static ProjectKey From(string projectFilePath)
+    // ProjectKey represents the path of the intermediate output path, which is where the project.razor.json file will
+    // end up. All creation logic is here in one place to ensure this is consistent.
+    public static ProjectKey From(HostProject hostProject) => new(hostProject.IntermediateOutputPath);
+    public static ProjectKey From(IProjectSnapshot project) => new(project.IntermediateOutputPath);
+    public static ProjectKey? From(Project project)
     {
-        // Right now, we expect this to be a project file path. This will change in future
-        Debug.Assert(projectFilePath.EndsWith(".csproj", StringComparison.OrdinalIgnoreCase));
-        return new(projectFilePath);
-    }
-    public static ProjectKey FromLegacy(string projectFilePath)
-    {
-        // In the legacy editor, we expect this to be a project file path.
-        Debug.Assert(projectFilePath.EndsWith(".csproj", StringComparison.OrdinalIgnoreCase));
-        return new(projectFilePath);
+        var intermediateOutputPath = Path.GetDirectoryName(project.CompilationOutputInfo.AssemblyPath);
+        return intermediateOutputPath is null ? null : new(intermediateOutputPath);
     }
 
     public string Id { get; }
 
     private ProjectKey(string id)
     {
         Debug.Assert(id is not null, "Cannot create a key for null Id. Did you call ProjectKey.From(this) in a constructor, before initializing a property?");
+        Debug.Assert(!id!.EndsWith(".csproj", StringComparison.OrdinalIgnoreCase), "We expect the intermediate output path, not the project file");
         // The null check in the assert means the compiler thinks we're lying about id being non-nullable.. which is fair I suppose.
-        Id = id.AssumeNotNull();
+        Id = FilePathNormalizer.NormalizeDirectory(id).AssumeNotNull();
```

Comment:
I'm running out of fumes at the end of the day, so I will be lazy and put the burden on you: is there a test that validates projects with not normalized intermediate paths all get the same id? If not, please add :)

Reply:
There are, but only via existing tests that fail without this line. Good idea to add a specific one though.

Reply:
Added some tests, but they started to look a lot like tests of the `FilePathNormalizer`, which we already have, but let me know if there is anything you think I've missed.